eis_toolkit/prediction/weights_of_evidence.py

```diff
@@ -0,0 +1,255 @@
+    """Read raster data and handle NoData values."""
+
+    array = np.array(raster.read(1), dtype=np.float32)
+
+    if nodata is not None:
+        array[array == nodata] = np.nan
+    elif raster.meta["nodata"] is not None:
+        array[array == raster.meta["nodata"]] = np.nan
+
+    return array
+
+
+def _calculate_metrics_for_class(deposits: np.ndarray, evidence: np.ndarray) -> tuple:
+    """Calculate weights/metrics for given data."""
+    A = np.sum(np.logical_and(deposits == 1, evidence == 1))
+    B = np.sum(np.logical_and(deposits == 1, evidence == 0))
+    C = np.sum(np.logical_and(deposits == 0, evidence == 1))
+    D = np.sum(np.logical_and(deposits == 0, evidence == 0))
+
+    if A + B == 0:
+        raise Exception("No deposits")
+    if C + D == 0:
+        raise Exception("All included cells have deposits")
+
+    if A == 0:
+        return A, B, C, D, 0, 0, 0, 0, 0, 0, 0
+
+    p_A_nominator = A
+    p_C_nominator = C
+    B_adjusted = B
+    D_adjusted = D
+
+    if B == 0:
+        p_A_nominator -= 0.99
+        B_adjusted = 0.99
+
+    if D == 0:
+        p_C_nominator -= 0.99
+        D_adjusted = 0.99
+
+    p_A = p_A_nominator / (A + B)  # probability of presence of evidence given the presence of mineral deposit
+    p_C = p_C_nominator / (C + D)  # probability of presence of evidence given the absence of mineral deposit
+
+    # Calculate metrics
+    w_plus = np.log(p_A / p_C) if p_C != 0 else 0  # Check
+    w_minus = np.log((1 - p_A) / (1 - p_C))
+    contrast = w_plus - w_minus
+
+    # Calculate signifigance metrics
+    s_w_plus = np.sqrt((1 / p_A_nominator) + (1 / p_C_nominator))
+    s_w_minus = np.sqrt((1 / B_adjusted) + (1 / D_adjusted))
+
+    s_contrast = np.sqrt(s_w_plus**2 + s_w_minus**2)
+    studentized_contrast = contrast / s_contrast
+
+    return A, B, C, D, w_plus, s_w_plus, w_minus, s_w_minus, contrast, s_contrast, studentized_contrast
+
+
+def _unique_weights(deposits: np.ndarray, evidence: np.ndarray) -> dict:
+    """Calculate unique weights for each class."""
+    classes = np.unique(evidence)
+    return {cls: _calculate_metrics_for_class(deposits, evidence == cls) for cls in classes}
+
+
+def _cumulative_weights(deposits: np.ndarray, evidence: np.ndarray, ascending: bool = True) -> dict:
+    """Calculate cumulative weights (ascending or descending) for each class."""
+    classes = sorted(np.unique(evidence), reverse=not ascending)
+    cumulative_classes = [classes[: i + 1] for i in range(len(classes))]
+    return {
+        cls[i]: _calculate_metrics_for_class(deposits, np.isin(evidence, cls))
+        for i, cls in enumerate(cumulative_classes)
+    }
+
+
+def _reclassify_by_studentized_contrast(df: pd.DataFrame, studentized_contrast_threshold: Number) -> None:
+    """Create generalized classes based on the studentized contrast threhsold value."""
+    index = df.idxmax()["Contrast"]
```

Comment:
```suggestion
    index = df.idxmax()[CONTRAST_COLUMN]
```

There is a lot of use of column name strings. There is a high chance for errors when you have to write the string each time explicitly. I would suggest creating static variables for each column name e.g. `CONTRAST_COLUMN = "Contrast"` and then using these in all the different functions.